docs/adrs/0549-composite-run-steps.md

```diff
@@ -0,0 +1,275 @@
+
+Moreover, the output ids are only accessible within the scope where it was defined. Note that in the example above, in our `workflow.yml` file, it should not have access to output id (i.e. `my-output`). For example, in the `workflow.yml` file, you can't run `foo.steps.my-step.my-output`.
+
+Our philosophy is we don't want the workflow author to need to know how the internal workings of the action work.
+
+### Context
+Similar to the workflow file, the composite action has access to the [same context objects](https://help.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#contexts) (ex: `github`, `env`, `strategy`). 
+
+### Environment
+Example `user/composite/action.yml`:
+```
+using: 'composite' 
+env:
+  NAME2: test2
+  SERVER: development
+steps: 
+  - id: my-step
+    run: |
+      echo NAME2 $NAME2
+      echo Server $SERVER 
+    env:
+      NAME2: test3
+```
+
+Example `workflow.yml`:
+```
+env:
+  NAME1: test1
+  SERVER: production
+steps: 
+  - id: foo
+    uses: user/test@v1
+  - run: echo Server $SERVER
+```
+
+Example Output:
+```
+NAME2 test3
+Server development
+Server production
+```
+
+We plan to use environment variables for Composite Actions similar to the parent/child relationship between nested function calls in programming languages like Python in terms of [lexical scoping](https://inst.eecs.berkeley.edu/~cs61a/fa19/assets/slides/29-Tail_Calls_full.pdf). In Python, let's say you have `functionA` that has local variables called `a` and `b` in this function frame. Let's say we have a `functionB` whose parent frame is `functionA` and has local variable `a` (aka `functionB` is called and defined in `functionA`). `functionB` will have access to its parent input variables that are not overwritten in the local scope (`a`) as well as its own local variable `b`. [Visual Example](http://www.pythontutor.com/visualize.html#code=def%20functionA%28%29%3A%0A%20%20%20%20a%20%3D%201%0A%20%20%20%20b%20%3D%202%0A%20%20%20%20def%20functionB%28%29%3A%0A%20%20%20%20%20%20%20%20b%20%3D%203%0A%20%20%20%20%20%20%20%20print%28%22a%22,%20a%29%0A%20%20%20%20%20%20%20%20print%28%22b%22,%20b%29%0A%20%20%20%20%20%20%20%20return%20b%0A%20%20%20%20return%20functionB%28%29%0A%0A%0A%0AfunctionA%28%29&cumulative=false&curInstr=14&heapPrimitives=nevernest&mode=display&origin=opt-frontend.js&py=3&rawInputLstJSON=%5B%5D&textReferences=false) 
+
+Similar to the above logic, the environment variables will flow from the parent node to its children node. More concretely, whatever workflow/action calls a composite action, that composite action has access to whatever environment variables its caller workflow/action has. Note that the composite action can append its own environment variables or overwrite its parent's environment variables. 
+
+
+### If Condition
+
+Example `user/composite/action.yml`:
```

Comment:
we may want to explicitly punt on this for now and move this to a separate ADR or appendix

Reply:
same for `continue-on-error`, `timeout-minutes`

Reply:
Agreed, I'd like the idea of starting with a more barebones implementation, and letting user feedback and scenarios drive how we can improve it going forward!